doc/source/development/rfc/rfc101_raster_dataset_threadsafety.rst

```diff
@@ -0,0 +1,474 @@
+
+.. code-block:: shell
+
+    $ time multireadtest -t 4 -i 1000000 20x20.tif
+    real    0m2.084s
+    user    0m8.155s
+    sys     0m0.020s
+
+    vs
+
+    $ time multireadtest -thread_safe -t 4 -i 1000000 20x20.tif
+    real    0m2.387s
+    user    0m9.334s
+    sys     0m0.029s
+
+
+But on a 4096x4096 raster with a number of iterations reduced to 100. the
+thread_safe mode is slightly faster:
+
+.. code-block:: shell
+
+    $ time multireadtest -t 4 -i 100 4096x4096.tif
+    real    0m4.845s
+    user    0m18.666s
+    sys     0m0.084s
+
+    vs
+
+    $ time multireadtest -thread_safe -t 4 -i 100 4096x4096.tif
+    real    0m4.255s
+    user    0m16.370s
+    sys     0m0.096s
+
+
+A Python equivalent of multireadtest has been written. Scalability depends
+on how much Python code is executed. If relatively few long-enough calls to
+GDAL are done, scalability tends to be good due to the Python Global Interpreter
+Lock (GIL) being dropped around them. If many short calls are done, the GIL
+itself, or its repeated acquisition and release, becomes the bottleneck. This is
+no different than using a GDALDataset per thread.
+
+Documentation
+-------------
+
+Documentation for the new constant and functions will be added. The
+:ref:`multithreading` page will be updated to reflect the new capability
+introduced by this RFC.
+
+Backward compatibility
+----------------------
+
+No issue anticipated: the C and C++ API are extended.
+The C++ ABI is modified due to additions of new virtual methods.
+
+Testing
+-------
+
+Tests will be added for the new functionality, including stress tests to have
+sufficiently high confidence in the correctness of the implementation for common
+use cases.
+
+Risks
+-----
+
+Like all code related to multi-threading, the C++ language and tooling offers
+hardly any safety belt against thread-safety programming errors. So it cannot
+be excluded that the implementation suffers from  bugs in some edge scenarios,
+or in the usage of some methods of GDALDataset, GDALRasterBand and related objects
+(particularly existing non-virtual methods of those classes that could happen
+to have a non thread-safe implementation)
+
+Design discussion
+-----------------
+
+This paragraph discusses a number of thoughts that arose during the writing of
+this RFC.
+
+1.  A significantly different alternative could have consisted in adding native
+    thread-safety in each driver. But this is not realistic for the following reasons:
+
+    * if that was feasible, it would require considerable development effort to
+      rework each drivers. So realistically, only a few select drivers would be updated.
+
+    * Even updating a reduced number of drivers would be extremely difficult, in
+      particular the GeoTIFF one, due to the underlying library not being reentrant,
+      and deferred loading strategies and many state variables being modified even
+      by read-only APIs. And this applies to most typical drivers.
```

Comment:
Would it be feasible for Zarr?

Reply:
> Would it be feasible for Zarr?

There are quite a number of state variables. For example opening a group or array adds entries to maps. Even reading pixels in a array involves some caches. So, likely much less difficult than GeoTIFF, but still not a trivial effort (and really hard to convince oneself that mutexes have been taken in all places where they should be)